js/background/background.js

```diff
@@ -674,88 +668,381 @@ class Steam {
+        
+        let expiry = await IndexedDB.db.get(objectStoreName, "expiry");
+        let expired;
+        if (!expiry) {
+            expired = true;
+        } else {
+            expired = IndexedDB.isExpired(expiry);
+        }
+        if (expired) {
+            await IndexedDB.clear(objectStoreName);
+            await IndexedDB.fetchUpdatedData(objectStoreName, null, params);
+        }
+    }
+
+    static async fetchUpdatedData(objectStoreName, key, params) {
+        if (!IndexedDB.cacheObjectStores.has(objectStoreName)) return;
+
+        let requestKey = key ? `${objectStoreName}_${key}` : objectStoreName;
+        if (IndexedDB._ongoingRequests.has(requestKey)) {
+            return IndexedDB._ongoingRequests.get(requestKey);
+        }
+
+        let req;
+        if (IndexedDB.timestampedObjectStores.has(objectStoreName)) {
+            req = IndexedDB.objStoreFetchFns.get(objectStoreName)(params);
+        } else {
+            req = IndexedDB.objStoreFetchFns.get(objectStoreName)(params, key);
+        }
+        req = req
+            .then(async () => {
+                if (key) {
+                    if (!await IndexedDB.db.transaction(objectStoreName).store.openKeyCursor(key)) {
+                        // Prevent fetching the same empty result for every db request for this key
+                        return IndexedDB.putCached(objectStoreName, null, key);
+                    }
+                }
+            })
+            .finally(() => IndexedDB._ongoingRequests.delete(requestKey));
+        IndexedDB._ongoingRequests.set(requestKey, req);
+        return req;        
+    }
+}
+IndexedDB._promise = null;
+IndexedDB._ongoingRequests = new Map();
+
+/*  Object stores in this map won't get checked
+    for timestamps if cached.
+    Instead of checking the single entry, the object store itself has
+    a entry named "expiry".
+
+    This allows us to reduce the overhead of having one timestamp for
+    each individual entry, although they're basically fetched during
+    the same time.
+*/
+IndexedDB.timestampedObjectStores = new Map([
+    ["coupons", 60 * 60],
+    ["giftsAndPasses", 60 * 60],
+    ["items", 60 * 60],
+    ["earlyAccessAppids", 60 * 60],
+    ["purchases", 24 * 60 * 60],
+    ["dynamicStore", 15 * 60],
+    ["ownedElsewhere", 15 * 60],
+    ["rates", 60 * 60],
+]);
+IndexedDB.cacheObjectStores = new Map([...IndexedDB.timestampedObjectStores,
+    ["packages", 7 * 24 * 60 * 60],
+    ["storePageData", 60 * 60],
+    ["profiles", 24 * 60 * 60],
+    ["itad", 15 * 60],
+]);
+
+// Functions that are called when an object store (or one of its entries) has expired
+IndexedDB.objStoreFetchFns = new Map([
+    ["coupons", SteamCommunity.coupons],
+    ["giftsAndPasses", SteamCommunity.giftsAndPasses],
+    ["items", SteamCommunity.items],
+    ["earlyAccessAppids", AugmentedSteamApi.endpointFactoryCached("v01/earlyaccess", "earlyAccessAppids", true, true)],
```

Comment:
Doesn't this make it execute the function? If it returns another function, it should be fine. I usually just use a quick arrow function if I wanna pass arguments without executing the actual function.

Reply:
Well, it returns another function, so that's intended